br/pkg/lightning/backend/external/byte_reader.go

```diff
@@ -54,16 +65,57 @@ func openStoreReaderAndSeek(
 
 // newByteReader wraps readNBytes functionality to storageReader. It will not
 // close storageReader when meet error.
-func newByteReader(ctx context.Context, storageReader storage.ReadSeekCloser, bufSize int) (*byteReader, error) {
+func newByteReader(ctx context.Context, storageReader storage.ReadSeekCloser, bufSize int, st storage.ExternalStorage, name string, defaultUseConcurrency bool) (*byteReader, error) {
+	conReader, err := NewSingeFileReader(ctx, st, name, 8, ConcurrentReaderBufferSize)
+	if err != nil {
+		return nil, err
+	}
 	r := &byteReader{
 		ctx:           ctx,
 		storageReader: storageReader,
 		buf:           make([]byte, bufSize),
 		bufOffset:     0,
+		conReader:     conReader,
 	}
+	r.SwitchReaderMode(defaultUseConcurrency)
 	return r, r.reload()
 }
 
+// SwitchReaderMode switches to concurrent reader.
+func (r *byteReader) SwitchReaderMode(useConcurrent bool) {
+	r.useConcurrentReader.Store(useConcurrent)
+}
+
+func (r *byteReader) switchToConcurrentReaderImpl() error {
+	if r.conReader == nil {
+		return errors.New("can't use the concurrent mode because reader is not initialized correctly")
+	}
+	currOffset, err := r.storageReader.Seek(0, io.SeekCurrent)
+	if err != nil {
+		return err
+	}
+	r.currFileOffset = currOffset
+	r.conReader.currentFileOffset = currOffset
+	r.conReader.bufferReadOffset = 0
+
+	r.useConcurrentReaderCurrent.Store(true)
+	r.conReader.buffer = make([]byte, r.conReader.concurrency*r.conReader.readBufferSize)
+	return nil
+}
+
+// UseConcurrentReader returns whether the reader is using concurrent reader.
+func (r *byteReader) UseConcurrentReader() bool {
```

Comment:
please add comment to explain which method of `byteReader` will be concurrently called with another one.

Reply:
I mean will `UseConcurrentReader` be called concurrently with `readNBytes` and other methods? Do we really need atomic variable?